pkg/tfvars/openstack/openstack.go

```diff
@@ -62,3 +73,69 @@ func TFVars(masterConfig *v1alpha1.OpenstackProviderSpec, cloud string, external
 
 	return json.MarshalIndent(cfg, "", "  ")
 }
+
+func getSwiftPublicURL(cloud string) (string, error) {
+	var swiftPublicURL string
+	serviceCatalog, err := getServiceCatalog(cloud)
+	if err != nil {
+		return "", err
+	}
+
+	for _, svc := range serviceCatalog.Entries {
+		if svc.Type == "object-store" {
+			for _, e := range svc.Endpoints {
+				if e.Interface == "public" {
+					swiftPublicURL = e.URL
+					break
+				}
+			}
+			break
+		}
+	}
+
+	if swiftPublicURL == "" {
+		return "", fmt.Errorf("cannot retrieve Swift URL from the service catalog")
+	}
+
+	return swiftPublicURL, nil
+}
+
+func getServiceCatalog(cloud string) (*tokens.ServiceCatalog, error) {
+	opts := &clientconfig.ClientOpts{
+		Cloud: cloud,
+	}
+
+	conn, err := clientconfig.NewServiceClient("identity", opts)
+	if err != nil {
+		return nil, err
+	}
+
+	cloudConfig, err := clientconfig.GetCloudFromYAML(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	domainName := cloudConfig.AuthInfo.DomainName
+	if domainName == "" {
+		domainName = cloudConfig.AuthInfo.UserDomainName
+	}
+
+	scope := tokens.Scope{
+		ProjectName: cloudConfig.AuthInfo.ProjectName,
+		DomainName:  domainName,
+	}
+
+	authOptions := tokens.AuthOptions{
+		Scope:      scope,
+		Username:   cloudConfig.AuthInfo.Username,
+		Password:   cloudConfig.AuthInfo.Password,
+		DomainName: domainName,
+	}
+
+	serviceCatalog, err := tokens.Create(conn, &authOptions).ExtractServiceCatalog()
```

Comment:
Hmm, i'm not sure what this does ? but does it create some resource that needs to be cleaned up on destroy?

Reply:
Here we want to obtain Swift public endpoint... By design this should be done by using https://www.terraform.io/docs/providers/openstack/d/identity_endpoint_v3.html but OpenStack default policies forbid to use this API for regular users.
On the other hand when you authenticate in OpenStack (i.e. get a token) it includes the whole service catalog in the output json. So we are able to parse it and get the endpoint from there https://docs.openstack.org/api-ref/identity/v3/?expanded=token-authentication-with-scoped-authorization-detail#token-authentication-with-scoped-authorization
In short the algorithm is:
1. Authenticate in OpenStack: `tokens.Create(..)`
2. Parse the token and extract the service catalog: `ExtractServiceCatalog()`
3. Iterate through the catalog and find `public` endpoint for `object-store`.

Unfortunately this feature is not supported by Terraform so I had to implement it here. No resources are created we just get endpoint data from OpenStack.

Btw, we do the same thing for Octavia service: https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/openstack/openstack.go#L141-L152